### DISCUSSION TITLE
When to use ptr in WGSL function parameters?

### DISCUSSION THREAD

**Discussion by alienself** [gpuweb/gpuweb#4830] [Q&A] [answered]
Hi,

Quick question, I would like to know when should I use ptr in my WGSL function parameters to avoid copy?

From what I understand by default all parameters are copied (not passed by reference).

Therefore, should I use ptr everywhere for anything bigger than 4 bytes (ptr size) OR only for structs? 

For instance a vec4<f32> is 16 bytes, is it a good candidate for ptr? What about matrix4 (64 bytes)?

**Comment by alienself** [comment]
anyone? :(

**Reply by kainino0x** [reply, marked as answer]
https://matrix.to/#/!MFogdGJfnZLrDmgkBN:matrix.org/$Gxx94gHIurZMAxzFOWLto37frl2bZlQ7fvYZrKuhUu4?via=matrix.org&via=mozilla.org&via=beeper.com

kainino (he/they)
> bumping a wgsl question maybe someone can help with
> https://github.com/gpuweb/gpuweb/discussions/4830

jasperrlz
> "Copy" doesn't really mean anything, because copies tend to be elided if they can be.

miketuritzin
> It's tricky because it's not obvious when copies are elided. I ran into a huge perf issue when I tried to pass a shared-mem array to a function without using ptr, only later to discover through some trial and error what the issue was.

cwfitzgerald
> yeah it's really easy to fall of that performance cliff

**Reply by alienself** [reply]
Thanks a lot @kainino0x! This was super useful. I have joined matrix so i will ask questions directly there next time.
